### PR TITLE
WT-3144 bug fix: random cursor returns not-found when descending to an empty page

### PR DESCRIPTION
**Pull request by keithbostic**
If random descent through the tree fails, fallback to skipping through the tree's pages; if skipping through the tree's pages fails, fallback to a random entry from the first page in the tree that contains anything at all.

### CODE REVIEW

File: src/btree/bt_cursor.c [hunk 1]
```diff
@@ -896,24 +896,53 @@ __wt_btcur_next_random(WT_CURSOR_BTREE *cbt)
...
 		    session, ret = __wt_row_random_descent(session, cbt));
+
+		/*
+		 * Random descent may return not-found: the tree might be empty
+		 * or have so many deleted items we didn't find any valid pages.
+		 * We can't return WT_NOTFOUND to the application unless a tree
+		 * is really empty, fallback to skipping through tree pages.
+		 */
+		if (ret == WT_NOTFOUND)
+			goto fallback_walk;
```
Comment (michaelcahill) [inline src/btree/bt_cursor.c]: As far as I can see, we won't have done the calculation of `cbt->next_random_leaf_skip`, so it looks like we'll always find the first record in the tree.  Am I missing something?

File: test/suite/test_cursor_random.py [hunk 2]
```diff
@@ -71,6 +71,15 @@ def test_cursor_random(self):
...
 
+    # Check that next_random fails with an empty tree, repeatedly.
+    def test_cursor_random_empty(self):
+        uri = self.type
+        self.session.create(uri, 'key_format=S,value_format=S')
+        cursor = self.session.open_cursor(uri, None, self.config)
+        for i in range(1,5):
+            self.assertTrue(cursor.next(), wiredtiger.WT_NOTFOUND)
+        cursor.close
+
```
Comment (michaelcahill) [inline test/suite/test_cursor_random.py]: It looks like we should have a test that creates a tree with enough data for multiple pages, reopens the connection so we have a real tree, then truncates most / all of the tree and makes sure random lookups find data / fail (respectively).  That way we're testing {{WT_REF_DELETED}}, not just empty pages.

File: src/btree/bt_cursor.c (hunk 1 above)
Comment (keithbostic) [inline src/btree/bt_cursor.c]: > we won't have done the calculation of cbt->next_random_leaf_skip, so it looks like we'll always find the first record in the tree.

Agreed, I've done another run at this one, ready for your review.

File: test/suite/test_cursor_random.py (hunk 2 above)
Comment (keithbostic) [inline test/suite/test_cursor_random.py]: > It looks like we should have a test that creates a tree with enough data for multiple pages, reopens the connection so we have a real tree, then truncates most / all of the tree and makes sure random lookups find data / fail (respectively). That way we're testing WT_REF_DELETED, not just empty pages.

Added in fc30d0e.

**Comment by michaelcahill** [comment]
Thanks @keithbostic, lgtm.  I'll merge.

**Pull request merged**